### PR TITLE
Feature/filter on region

### PR DESCRIPTION
**Pull request by PinkhamSoftware**
What:
- Search on Region
- Get Searchable regions as a List (for purpose of populating dropdown list).

Why:
- Allow finance team to filter by region and gain an understanding of investments via broad geographical area.

Notes:
- Only pulls list of regions from LocationRegionName, which appears to be the newer region name.
- Allows to search on both the older or newer region name.

### CODE REVIEW

File: WebApi/Controllers/Search/SearchFiltersController.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+using System.Threading.Tasks;
+using HomesEngland.Domain;
+using HomesEngland.Gateway.AssetRegisterVersions;
+using HomesEngland.UseCase.CalculateAssetAggregates.Models;
+using Microsoft.AspNetCore.Mvc;
+using WebApi.Extensions;
+
+namespace WebApi.Controllers.Search
+{
+    [Route("api/v1/asset/search/filters")]
```
Comment (danielburnley) [inline WebApi/Controllers/Search/SearchFiltersController.cs]: Not sure this route is right - not sure what it would be though :thinking:

Comment (PinkhamSoftware) [reply WebApi/Controllers/Search/SearchFiltersController.cs]: Yep, I wasn't sure but I went with that, happy to change it.

File: AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+using HomesEngland.UseCase.SearchAsset.Models;
+using Main;
+using Microsoft.EntityFrameworkCore;
+using NUnit.Framework;
+using TestHelper;
+
+namespace AssetRegisterTests.HomesEngland.UseCases.Search.Region
+{
+    [TestFixture]
+    public class GetAssetRegionsUseCaseAcceptanceTests
```
Comment (danielburnley) [inline AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs]: So for this test - I'd actually look at involving it in the search acceptance test. 

Coming from the perspective of the user:
- Setup: Create a bunch of assets
- Get a list of all the regions
- Search by one of those regions found in the results
- Assert that the results match what you expect

This way you have acceptance tests build around the user of the API as opposed to lots of small steps that a user would be doing all as one, if that makes sense

File: HomesEngland.Gateway.Test/AssetGatewayTests.cs [hunk 3]
```diff
@@ -60,4 +61,59 @@ public async Task GivenMultipleAssetsHaveBeenCreated_WhenTheAssetsAreReadFromThe
             }
         }
     }
+
+    [TestFixture]
+    public class GetAssetRegionsGatewayTests
```
Comment (danielburnley) [inline HomesEngland.Gateway.Test/AssetGatewayTests.cs]: Feels like you could move the region list tests to their own file, as they're only testing that interface

Comment (PinkhamSoftware) [reply HomesEngland.Gateway.Test/AssetGatewayTests.cs]: Yep

File: AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs (hunk 2 above)
Comment (PinkhamSoftware) [reply AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs]: Ok yep good suggestion.

File: HomesEngland.Gateway/Sql/EFAssetGateway.cs [hunk 4]
```diff
@@ -85,6 +85,13 @@ private IQueryable<AssetEntity> GenerateFilteringCriteria(AssetRegisterContext c
                     EF.Functions.Like(w.Address.ToLower(), $"%{searchRequest.Address}%".ToLower()));
             }
 
+            if(!string.IsNullOrEmpty(searchRequest.Region) && !string.IsNullOrWhiteSpace(searchRequest.Region))
+            {
+                queryable = queryable.Where(w =>
+                    EF.Functions.Like(w.LocationLaRegionName.ToLower(), $"%{searchRequest.Region}%".ToLower()) || 
+                    EF.Functions.Like(w.ImsOldRegion.ToLower(), $"%{searchRequest.Region}%".ToLower()));
```
Comment (danielburnley) [inline HomesEngland.Gateway/Sql/EFAssetGateway.cs]: :thinking: Do we know which of these two regions is the one they'd want to filter on/if they can be different? Just thinking I don't personally know the difference between them so treating them as equals might not actually be the expected behaviour. Might be worth clarifying this before we merge this so we know for sure the expected behaviour

File: HomesEngland.Gateway/Sql/EFAssetGateway.cs [hunk 5]
```diff
@@ -122,5 +129,18 @@ public Task<IAssetAggregation> Aggregate(IAssetSearchQuery searchRequest, Cancel
...
             }
         }
+
+        public Task<IList<AssetRegion>> ListRegionsAsync(CancellationToken cancellationToken)
+        {
+            using (var context = new AssetRegisterContext(_databaseUrl))
+            {
+                IList<AssetRegion> results = context.Assets.Select(s => new AssetRegion
+                {
+                    Name = s.LocationLaRegionName
```
Comment (danielburnley) [inline HomesEngland.Gateway/Sql/EFAssetGateway.cs]: Interesting that you only list the `LocationLaRegionName` here but it allows searching on both - was there a reason for this?

File: HomesEngland.Gateway/Sql/EFAssetGateway.cs (hunk 4 above)
Comment (PinkhamSoftware) [reply HomesEngland.Gateway/Sql/EFAssetGateway.cs]: Yep, good point 👍

File: AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            foundAsset.Assets.Should().NotBeNullOrEmpty();
+            foundAsset.Assets.ElementAt(0).AssetOutputModelIsEqual(createdAsset.Asset);
+        }
+
+        private CreateAssetRequest CreateAsset( string region)
+        {
+            CreateAssetRequest createAssetRequest = TestData.UseCase.GenerateCreateAssetRequest();
+            if (!string.IsNullOrEmpty(region))
+            {
+                createAssetRequest.LocationLaRegionName = region;
```
Comment (danielburnley) [inline AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs]: Probably should remove this as it's not relevant to the test

Comment (PinkhamSoftware) [reply AcceptanceTest/HomesEngland/UseCases/Search/Region/GetAssetRegionsUseCaseAcceptanceTests.cs]: Ok I'll go and remove them.

**Pull request merged**